src/lib/snapp.ts

```diff
@@ -225,29 +278,50 @@ type ExecutionState = {
  * ```
  *
  */
-export abstract class SmartContract {
-  // protocolState: ProtocolStatePredicate;
+export class SmartContract {
   address: PublicKey;
 
-  // state: Array<State<Field>>;
-
-  // private _self: { body: Body, predicate: AccountPredicate } | undefined;
-
-  _executionState: ExecutionState | undefined;
+  _executionState?: ExecutionState;
+  static _methods?: methodEntry<SmartContract>[];
 
   constructor(address: PublicKey) {
     this.address = address;
-    // this.self = null as unknown as Body;
-    // this.state = [];
+  }
+
+  static compile(address?: PublicKey) {
+    // TODO: think about how address should be passed in
+    // if address is not provided, create a random one
+    address ??= PrivateKey.random().toPublicKey();
+    let dummyInstance = new this(address);
+    let inductiveRules = (this._methods ?? []).map(
+      ({ methodName, witnessArgs }) => {
+        function main(transactionHash: Field) {
+          Mina.setCurrentTransaction({
+            sender: PrivateKey.random(), // TODO
+            parties: [],
+            nextPartyIndex: 0,
+            protocolState: ProtocolStatePredicate.ignoreAll(),
+          });
+          let witnesses = witnessArgs.map(emptyWitness); // TODO is this correct?
+          (dummyInstance[methodName] as any)(...witnesses); // call method in a way that `this` is defined
+          let hash = computeTransactionHash(Mina.currentTransaction);
+          console.log({ transactionHash, hash });
+          hash.assertEquals(transactionHash);
+          Mina.setCurrentTransaction(undefined);
+        }
+        return [0, methodName, main];
+      }
+    );
+    let output = picklesCompile(inductiveRules);
+    console.log('output', output);
+    return output;
   }
 
   deploy(...args: any[]) {
     try {
       this.executionState().party.body.update.verificationKey.set = Bool(true);
-    } catch (_error) {
-      throw new Error(
-        'Cannot deploy SmartContract outside a transaction.'
-      );
+    } catch {
+      throw new Error('Cannot deploy SmartContract outside a transaction.');
```

Comment:
Do these errors get surfaced up to the user? Maybe we can add an additional text snippet alongside this message that helpfully guides the user on how to fix this if there is a common fix. For this, something like `Make sure you are using Mina.transaction` ? Wondering what everyone else thinks.

Reply:
Good idea!

Reply:
In the current version, this particular error is actually not present any more, but a very similar error mentioning `Mina.transaction` would be thrown: https://github.com/o1-labs/snarkyjs/blob/f0014217b9911438da621751a33f7bd3c97f3586/src/lib/zkapp.ts#L485

Reply:
See #130 and my other comment